src/librbd/internal.cc

```diff
@@ -1635,6 +1636,228 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
+          return -EBUSY;
+        }
+      }
+    }
+
+    BOOST_SCOPE_EXIT_ALL(ictx, cct) {
+      bool is_locked = ictx->exclusive_lock != nullptr &&
+                       ictx->exclusive_lock->is_lock_owner();
+      ictx->owner_lock.put_read();
+      if (is_locked) {
+        C_SaferCond ctx;
+        ictx->exclusive_lock->shut_down(&ctx);
+        int r = ctx.wait();
+        if (r < 0) {
+          lderr(cct) << "error shutting down exclusive lock" << dendl;
+        }
+      }
+      ictx->state->close();
+    };
+
+    ldout(cct, 2) << "adding image entry to rbd_trash" << dendl;
+    utime_t etp = ceph_clock_now();
+    utime_t ts = etp;
+    ts += (double)delay;
+    cls::rbd::DeletedImageSource del_source;
+    switch (source) {
+      case RBD_USER_TRASH_MOVE:
+        del_source = cls::rbd::USER;
+        break;
+      case RBD_RBD_MIRROR_TRASH_MOVE:
+        del_source = cls::rbd::RBD_MIRROR;
+        break;
+      default:
+        del_source = cls::rbd::USER;
+    }
+    cls::rbd::DeletedImageSpec delete_spec(del_source, image_name, ts, etp);
+    r = cls_client::trash_add(&io_ctx, image_id, delete_spec);
+    if (r < 0 && r != -EEXIST) {
+      lderr(cct) << "error adding image " << image_name << " to rbd_trash"
+                 << dendl;
+      return r;
+    } else if (r == -EEXIST) {
+      ldout(cct, 10) << "found previous unfinished deferred remove for image:"
+                     << image_id << dendl;
+      // continue with removing image from directory
+    }
+
+    ldout(cct, 2) << "removing id object..." << dendl;
+    r = io_ctx.remove(util::id_obj_name(image_name));
+    if (r < 0 && r != -ENOENT) {
+      lderr(cct) << "error removing id object: " << cpp_strerror(r)
+                 << dendl;
+      return r;
+    }
+
+    ldout(cct, 2) << "removing rbd image from v2 directory..." << dendl;
+    r = cls_client::dir_remove_image(&io_ctx, RBD_DIRECTORY, image_name,
+                                     image_id);
+    if (r < 0) {
+      if (r != -ENOENT) {
+        lderr(cct) << "error removing image from v2 directory: "
+                   << cpp_strerror(-r) << dendl;
+      }
+      return r;
+    }
+
+    return 0;
+  }
+
+  int trash_list(IoCtx &io_ctx, map<string, trash_info_t> &entries) {
+    CephContext *cct((CephContext *)io_ctx.cct());
+    ldout(cct, 20) << "trash_list " << &io_ctx << dendl;
+
+    map<string, cls::rbd::DeletedImageSpec> trash_entries;
+    int r = cls_client::trash_list(&io_ctx,  &trash_entries);
+    if (r < 0) {
+      lderr(cct) << "error listing rbd_trash entries" << dendl;
+      return r;
+    }
+
+    for (const auto &entry : trash_entries) {
+      rbd_trash_move_source_t source;
+      switch (entry.second.source) {
+        case cls::rbd::USER:
+          source = RBD_USER_TRASH_MOVE;
+          break;
+        case cls::rbd::RBD_MIRROR:
+          source = RBD_RBD_MIRROR_TRASH_MOVE;
+          break;
+        default:
+          source = RBD_USER_TRASH_MOVE;
```

Comment:
Nit: I'd say either treat this as an error or just reinterpret_cast between the two enums